File: 01_Easy/0118_Pascals_Triangle.cpp
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> generate(int numRows) {
        // Base case
        if (numRows == 1){
            return vector<vector<int>> {{1}};
        }

        // ------------- Recursion Steps ----------------
        // Get the previous triangle
        vector<vector<int>> triangle = this->generate(numRows - 1);

        // Get the last generated row
        vector<int> previous_row = triangle.at(numRows - 2);

        // Generate the next row based on the previous row
        vector<int> next_row(numRows, 1);
        for (int i=1; i < numRows-1; i++){
            next_row.at(i) = previous_row.at(i-1) + previous_row.at(i);
        }

        // Add it to the triangle
        triangle.insert(triangle.end(), next_row);

        // Return the triangle
        return triangle;
    }
};

File: 01_Easy/0118_Pascals_Triangle.py
class Solution:
    def generate(self, numRows: int) -> list[list[int]]:
        # Base case
        if numRows == 1:
            return [[1]]

        # Get the previous row
        triangle = self.generate(numRows-1)
        previous_row = triangle[-1]

        # Build the current row based on the previous one
        current_row = [1] * numRows
        for i in range(1,numRows-1):
            current_row[i] = previous_row[i-1] + previous_row[i]
        
        triangle.append(current_row)
        return triangle
        
            